src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala

```diff
@@ -1,41 +1,95 @@
+      If infinite loop (and eventually StackOverflow) occurs when adding header/footer, check that the
+      page size is large enough/page margin is small enough so that header/footer content do not exceed
+      the available space in the document.
+   */
+  final val PAGE_MARGIN = 15
+  final val font = new Font(Font.HELVETICA, 12f, Font.UNDERLINE, Color.BLUE)
+  final val TABLE_ROW_HEIGHT = 23f
+
+
   private def generatePageSize(node: ElkNode): (Float, Float) = {
-    val width = node.getWidth.toFloat + 2 * ElkNodePainter.margin.toFloat
-    val height = node.getHeight.toFloat + 2 * ElkNodePainter.margin.toFloat
+    val width = node.getWidth.toFloat + 2.5f * ElkNodePainter.margin.toFloat
+    val height = node.getHeight.toFloat + 2.5f * ElkNodePainter.margin.toFloat
     (width, height)
   }
 
+
+  def getDuplicationList(block: HierarchyBlock, path: DesignPath = DesignPath(),
+                         dupList: mutable.Map[LibraryPath, Set[DesignPath]]): Unit ={
+
+    val className = block.getSelfClass
+    dupList.getOrElseUpdate(className, Set(path)) match {
+      case existingSet: Set[DesignPath] => dupList(className) = existingSet + path
+    }
+
+    block.blocks.asPairs.map {
+      case (name, subblock) => (name, subblock.`type`)
+    }.collect {
+      case (name, BlockLike.Type.Hierarchy(subblock)) =>  (path + name, subblock)
+    }.foreach {
+      case (path, subblock) => getDuplicationList(subblock, path, dupList)
+    }
+  }
+
+
   def generate(content: HierarchyBlock,
                mappers: Seq[PropertyMapper[NodeDataWrapper, PortWrapper, EdgeWrapper]] = Seq(),
                fileName: String): Unit = {
+
+    val dupList = mutable.Map[LibraryPath, Set[DesignPath]]()
+    getDuplicationList(content, dupList = dupList)
+
     try {
       val document = new Document()
       val writer = PdfWriter.getInstance(document, new FileOutputStream(fileName))
-      val footer = new HeaderFooter(true)
-      footer.setBorder(Rectangle.NO_BORDER)
-      footer.setAlignment(Element.ALIGN_RIGHT)
-      document.setFooter(footer)
 
-      def printNode(node: ElkNode): Unit = {
+
+      def printNode(node: ElkNode, className: LibraryPath, path: DesignPath): Unit = {
+        val dupSet = dupList.getOrElse(className, Set.empty)
         val (width, height) = generatePageSize(node)
-        document.setPageSize(new Rectangle(width, height))
+        val adjustedHeight = if(dupSet.size == 1) {
+          height
+        } else {
+          TABLE_ROW_HEIGHT * dupSet.size + height + ElkNodePainter.margin
+        }
+        document.setPageSize(new Rectangle(width, adjustedHeight))
+        document.setMargins(2*PAGE_MARGIN, 2*PAGE_MARGIN, PAGE_MARGIN, PAGE_MARGIN)
+
+        // Sets the header and footer of each page. Target of the component target link is set as the header.
+        val targetAnchor = new Anchor(path.toString)
+        targetAnchor.setName(path.toString)
+        val headerPhrase = new Phrase("Design Path: ")
```

Comment:
Would it be possible to eliminate this (so it just shows the path)?

We can also eliminate the (redundant) path from the block diagram.

Maybe also move the class name from the block diagram into the header, since it's creating overlap issues in some cases? (could be a follow-up PR)

Reply:
1. Do you mean eliminating the `headerPhrase`?

2 & 3. these refer to modifying the drawing algorithm to not print the path/class name from the block diagram right? I can look into it.

Reply:
1) Yeah, eliminating the "Design Path: " - effectively reduce visual noise

2/3) Yeah, that would require a change to the drawing algorithm and affect the block diagram viewer. But I also want to change that too, because the overlap is a problem.